Library/Homebrew/utils/fork.rb

```diff
@@ -33,10 +44,27 @@ def self.safe_fork(&_block)
             socket.close
           end
           write.close
-          data = read.read
+          # Each line on the error pipe contains a JSON-serialized exception.
+          # We read the first, since only one is necessary for a failure.
+          data = read.gets
           read.close
           Process.wait(pid) unless socket.nil?
-          raise Marshal.load(data) unless data.nil? || data.empty? # rubocop:disable Security/MarshalLoad
+
+          if data && !data.empty?
+            error_hash = JSON.parse(data) unless data.nil? || data.empty?
+
+            e = case error_hash["json_class"]
+            when "ErrorDuringExecution"
+              ErrorDuringExecution.new(error_hash["cmd"],
+                                       status: error_hash["status"],
+                                       output: error_hash["output"])
+            else
+              ChildProcessError.new(error_hash)
```

Comment:
Could this be caught and handled somewhere elsewhere so the fact it's come from a child process doesn't need to be handled in each and every case above?

Reply:
Where else were you thinking? I'm not sure where else to put it, since `fork.rb` is where the parent and child processes converge.

Reply:
Hmm, I could do something like this:

```ruby
e = ChildProcessError.new(error_hash)

raise e unless e.inner_class == ErrorDuringExecution

# Rebuild and throw the ErrorDuringExecution here
```

That flattens the logic a bit and gets rid of the ugly string comparison above, but it also creates an extra `ChildProcessError` that we never throw.

Reply:
@woodruffw Yeh, something like that would be good, I think. It'd be good to recreate the original exception and throw it as "early" as possible so that things that are "later"/higher up the call stack can avoid needing to know or care about `ChildProcessError` being a thing. It's possible I've misunderstood here, though, so please feel free to argue and disagree 😆